### DISCUSSION TITLE
I2C Issue

### DISCUSSION THREAD

**Discussion by A622266** [micropython/micropython#9727] [Hardware & Peripherals] [answered]
I'm using an [ESP32 S3 dev kit](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/hw-reference/esp32s3/user-guide-devkitc-1.html).  I am trying to detect another device on I2C:
```py
from machine import I2C, Pin

i2c = I2C(0, Pin(0), Pin(1),freq=400000)
print(i2c)
print(i2c.scan()) # correct result is 56, or 0x38
```
This code results in an error:
```
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
TypeError: extra positional arguments given
```

Is my code not following the I2C prototype? 

If I try to use the defaults:
```
from machine import I2C, Pin

i2c = I2C(0)
print(i2c)
print(i2c.scan()) # correct result is 56, or 0x38
```
This does run, and it produces this output:
```
I2C(0, scl=0, sda=1, freq=851063)
[]
```
I expected to see ```scl=18``` and ```sda=19```, as those are the default pins for the hardware I2C according to the quick reference for the esp32 (although I have an S3 dev kit, so maybe the defaults are not the same as esp32).  Also, the frequency parameter is not expected.  Can anyone give advice to configure I2C correctly using this dev kit?

**Comment by robert-hh** [comment, marked as answer]
Please have a look at the example in the documentation: https://docs.micropython.org/en/latest/esp32/quickref.html#hardware-i2c-bus
The template:
```
i2c = I2C(0)
i2c = I2C(1, scl=Pin(0), sda=Pin(1), freq=400000)
```
should be right and create output, if the wiring is for GND, 3.3V, SCL at Pin 0, SDA at Pin 1, and pull-up resistors are in place at SDA and SCL.

**Reply by A622266** [reply]
I believe I am meeting all the requirements you list, but I don't get the expected result.
With nothing connected to the esp32 board, and running the template:
```
from machine import I2C, Pin
i2c = I2C(0)
print(i2c)
i2c = I2C(1, scl=Pin(0), sda=Pin(1), freq=400000)
print(i2c)
print(i2c.scan())
```
I get this result:
```
I2C(0, scl=9, sda=8, freq=851063)
I2C(1, scl=0, sda=1, freq=851063)
[]
```
Does the output make sense (again, nothing connected to esp32 board)?
Shouldn't it be:
```
I2C(0, scl=18, sda=19, freq=400000)
I2C(1, scl=0, sda=1, freq=400000)
[]
```

**Reply by robert-hh** [reply]
The reported freq value is too high. Some peripherals may not work with that value. I recall that there is a PR to fix it. Bus besides that it looks OK. You can ty to set a lower freq value to get values <= 400000. The default pins for I2C change with different board types. So the report of Pin 8 and 9 does not look wrong.
If you have connected a device according to your script properly at pins 0 and 1, it should show up in scan().

**Comment by steveh127** [comment]
The ESP32 is very flexible, so long as you specify which pins are scl and sda you can use any available pins. And pull up resistors are essential though many I2C based modules include them. I routinely use ESP32 S3 kit with I2C with no problems. Do double check your wiring.
